nexus/src/app/mod.rs

```diff
@@ -190,20 +176,62 @@ pub struct Nexus {
     /// DNS resolver Nexus uses to resolve an external host
     external_resolver: Arc<external_dns::Resolver>,
 
-    /// DNS servers used in `external_resolver`, used to provide DNS servers to
-    /// instances via DHCP
-    // TODO: This needs to be moved to the database.
-    // https://github.com/oxidecomputer/omicron/issues/3732
-    external_dns_servers: Vec<IpAddr>,
-
     /// Background tasks
-    background_tasks: background::BackgroundTasks,
-
-    /// Default Crucible region allocation strategy
-    default_region_allocation_strategy: RegionAllocationStrategy,
-
-    /// Channel for notifying background task of change to opte v2p state
-    v2p_notification_tx: tokio::sync::watch::Sender<()>,
+    background_tasks: Arc<background::BackgroundTasks>,
+
+    /**
+     *
+     * Application level types that previously were methods on Nexus. These
+     * types are pub since they only provide methods, and adding getters seems
+     * superfluous. Each of these types is clonable for use in passing around
+     * the system.
+     *
+     * Please do not `impl super::Nexus` in any other files, and consider
+     * attaching new methods to distinct types.
+     *
+     * We may decide to merge some of these types, put them in external crates,
+     * etc.. Separating the methods into distinct types makes the above possible
+     * and also allows us to use them in background tasks where we can't cleanly
+     * pass Nexus due to the circular dependency.
+     *
+     **/
+    pub address_lot: address_lot::AddressLot,
+    pub source_ip_allow_list: allow_list::SourceIpAllowList,
+    pub bfd: bfd::Bfd,
+    pub bgp: bgp::Bgp,
+    pub certificate: certificate::Certificate,
+    pub crucible: crucible::Crucible,
+    pub blueprint: deployment::Blueprint,
+    pub device_auth: device_auth::DeviceAuth,
+    pub disk: disk::Disk,
+    pub external_ip: external_ip::ExternalIp,
+    pub iam: iam::Iam,
+    pub image: image::Image,
+    pub instance_network: instance_network::InstanceNetwork,
+    pub instance: instance::Instance,
+    pub ip_pool: ip_pool::IpPool,
+    pub metrics: metrics::Metrics,
+    pub network_interface: network_interface::NetworkInterface,
+    pub oximeter: oximeter::Oximeter,
+    pub probe: probe::Probe,
+    pub project: project::Project,
+    pub quota: quota::Quota,
+    pub rack: rack::Rack,
+    pub session: session::Session,
+    pub silo: silo::Silo,
+    pub sled: sled::Sled,
+    pub snapshot: snapshot::Snapshot,
+    pub ssh_key: ssh_key::SshKey,
+    pub switch_interface: switch_interface::SwitchInterface,
+    pub switch_port: switch_port::SwitchPort,
+    pub switch: switch::Switch,
+    pub update: update::Update,
+    pub utilization: utilization::Utilization,
+    pub volume: volume::Volume,
+    pub vpc_router: vpc_router::VpcRouter,
+    pub vpc_subnet: vpc_subnet::VpcSubnet,
+    pub vpc: vpc::Vpc,
+    pub saga_context: saga_interface::SagaContext,
```

Comment:
While I'm on-board with refactoring portions of code to only use dependencies that are actually necessary, this ends up carrying some structural / lifetime costs that seem non-ideal to me. Basically: For each module of code we add, we now require additional Arc clones, we increase the size of this structure indefinitely, etc.

Is that actually necessary? Is it important to make each one of these objects that have a lifetime tied to Nexus?

Let's say an operation arrives that wants to access something from the `SwitchInterface` submodule. Would it be possible to create a new `SwitchInterface` struct, containing *references* to the fields borrowed from `Nexus`, rather than cloning them? (E.g., grab a `&Arc<Datastore>`, rather than cloning?)

I think this is slightly closer to the approach @jgallagher took when creating https://github.com/oxidecomputer/omicron/tree/main/nexus/networking -- basically, pull out the functionality from Nexus, make it accessible to Nexus, but without creating a new object that "owns" these fields indefinitely. Even though free functions were used in that module, it seems possible to create on-the-fly "structs containing references to borrowed data" -- that seems effectively the same, as the lifetime of the borrowed fields would still be "in the big Nexus object".

Reply:
> While I'm on-board with refactoring portions of code to only use dependencies that are actually necessary, this ends up carrying some structural / lifetime costs that seem non-ideal to me. Basically: For each module of code we add, we now require additional Arc clones, we increase the size of this structure indefinitely, etc.
> 

I view this all as an intermediate state to get to where I want to go, which is figuring out which of these types can actually be merged into one type or more functional types, and then broken out into their own crates. I don't want to keep adding new types here, but the only way to make progress towards not having Nexus itself be required everywhere but the HTTP handlers was do this all in one shot. I definitely tried not doing it, as this was a lot of tedious work.

> Is that actually necessary? Is it important to make each one of these objects that have a lifetime tied to Nexus?
> 
> Let's say an operation arrives that wants to access something from the `SwitchInterface` submodule. Would it be possible to create a new `SwitchInterface` struct, containing _references_ to the fields borrowed from `Nexus`, rather than cloning them? (E.g., grab a `&Arc<Datastore>`, rather than cloning?)
> 

This may be possible to do in some cases, but I don't see how it works with async tokio tasks. Those all need static lifetimes on references as they can move across threads. This is especially problematic for background tasks.

> I think this is slightly closer to the approach @jgallagher took when creating https://github.com/oxidecomputer/omicron/tree/main/nexus/networking -- basically, pull out the functionality from Nexus, make it accessible to Nexus, but without creating a new object that "owns" these fields indefinitely. Even though free functions were used in that module, it seems possible to create on-the-fly "structs containing references to borrowed data" -- that seems effectively the same, as the lifetime of the borrowed fields would still be "in the big Nexus object".

I'm not particularly a fan of the free-function pattern, and based on lifetime issues, I don't see how we'd do what you are suggesting unfortunately.